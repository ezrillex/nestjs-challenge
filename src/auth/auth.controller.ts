import {
  Body,
  Controller,
  HttpCode,
  HttpStatus,
  Post,
  Request,
  UseGuards,
} from '@nestjs/common';
import { SignupUserDto } from './dto/signup-user.dto';
import { AuthService } from './auth.service';
import { LoginUserDto } from './dto/login-user.dto';
import { ForgotPasswordDto } from './dto/forgot-password.dto';
import { ResetPasswordDto } from './dto/reset-password.dto';
import { Public } from '../common/decorators/public.decorator';
import { Throttle, ThrottlerGuard } from '@nestjs/throttler';

@Controller('auth')
export class AuthController {
  constructor(private readonly authService: AuthService) {}

  @Public()
  @Post('signup')
  async signup(@Body() data: SignupUserDto) {
    const created_user = await this.authService.registerUser(data);
    return created_user;
  }

  @Public()
  @HttpCode(HttpStatus.OK)
  @Post('login')
  async login(@Body() data: LoginUserDto) {
    return this.authService.loginUser(data);
  }

  @HttpCode(HttpStatus.NO_CONTENT)
  @Post('logout')
  async logout(@Request() request: Request) {
    const userId = request['user'].id;
    return this.authService.logoutUser(userId);
  }

  @Public()
  @HttpCode(HttpStatus.OK)
  @Post('forgot')
  async forgot(@Body() data: ForgotPasswordDto) {
    return this.authService.forgotPassword(data);
  }

  @Public()
  @Throttle({ default: { limit: 3, ttl: 60000 } })
  @UseGuards(ThrottlerGuard)
  @HttpCode(HttpStatus.OK)
  @Post('reset')
  async reset(@Body() data: ResetPasswordDto) {
    return this.authService.changePassword(data);
  }
}
