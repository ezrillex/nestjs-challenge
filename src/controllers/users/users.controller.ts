import { Controller, Get, HttpCode, HttpStatus } from '@nestjs/common';

@Controller('users')
export class UsersController {
  @Get()
  @HttpCode(HttpStatus.OK)
  async getUsers() {
    return 'Hello World!';
  }
}
