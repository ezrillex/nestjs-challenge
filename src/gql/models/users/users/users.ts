import { Field, ID, ObjectType } from '@nestjs/graphql';
import { LikesOfProducts } from '../../products/likes/likes_of_products/likes_of_products';
import { IsEmail } from 'class-validator';

// partial implementation of actual table due to graphql not dealing with auth and roles
@ObjectType()
export class Users {
  @Field(() => ID)
  id: string;

  @Field(() => String)
  first_name: string;

  @Field(() => String)
  last_name: string;

  @IsEmail()
  @Field(() => String)
  email: string;

  @Field(() => [LikesOfProducts])
  likes_products: LikesOfProducts[];
}
