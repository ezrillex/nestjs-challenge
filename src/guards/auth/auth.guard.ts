import { CanActivate, ExecutionContext, Injectable } from '@nestjs/common';
import { Observable } from 'rxjs';
import { Reflector } from '@nestjs/core';
import { IS_PUBLIC_KEY } from '../../decorators/public/public.decorator';

@Injectable()
export class AuthGuard implements CanActivate {
  constructor(private reflector: Reflector) {}

  canActivate(
    context: ExecutionContext,
  ): boolean | Promise<boolean> | Observable<boolean> {
    const isPublic = this.reflector.getAllAndOverride<boolean>(IS_PUBLIC_KEY, [
      context.getHandler(),
      context.getClass(),
    ]);
    if (isPublic) {
      return true;
    }

    return false;
  }
}
