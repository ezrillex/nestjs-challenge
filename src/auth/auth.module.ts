import { Module } from '@nestjs/common';
import { JwtModule } from '@nestjs/jwt';
import { ConfigModule, ConfigService } from '@nestjs/config';
import { AuthController } from './auth.controller';
import { AuthService } from './auth.service';
import { PrismaService } from '../prisma/prisma.service';
import { APP_GUARD } from '@nestjs/core';
import { AuthGuard } from './guards/auth.guard';
import { RolesGuard } from './guards/roles.guard';
import { ThrottlerModule } from '@nestjs/throttler';
import { EmailsModule } from '../emails/emails.module';

@Module({
  controllers: [AuthController],
  providers: [
    AuthService,
    PrismaService,
    {
      provide: APP_GUARD,
      useClass: AuthGuard,
    },
    {
      provide: APP_GUARD,
      useClass: RolesGuard,
    },
  ],
  imports: [
    JwtModule.registerAsync({
      imports: [ConfigModule],
      global: true,
      useFactory: async (configService: ConfigService) => ({
        secret: configService.get<string>('JWTCONSTANT'),
        signOptions: { expiresIn: '24h' },
      }),
      inject: [ConfigService],
    }),
    ThrottlerModule.forRoot([{ ttl: 60000, limit: 100 }]),
    EmailsModule,
  ],
})
export class AuthModule {}
