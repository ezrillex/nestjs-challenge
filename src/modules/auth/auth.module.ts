import { Module } from '@nestjs/common';
import { JwtModule } from '@nestjs/jwt';
import { ConfigModule, ConfigService } from '@nestjs/config';
import { AuthController } from '../../controllers/authentication/auth.controller';
import { AuthService } from '../../services/authentication/auth.service';
import { PrismaService } from '../../services/prisma/prisma.service';

@Module({
  controllers: [AuthController],
  providers: [AuthService, PrismaService],
  imports: [JwtModule.registerAsync({
    imports: [ConfigModule],
    global: true,
    useFactory: async (configService: ConfigService) => ({
      secret: configService.get<string>('JWTCONSTANT'),
      signOptions: { expiresIn: '24h' },
    }),
    inject: [ConfigService],
  })]
})
export class AuthModule {}
