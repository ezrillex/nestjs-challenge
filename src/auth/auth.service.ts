import { HttpException, HttpStatus, Injectable } from '@nestjs/common';
import { SignupUserDto } from './dto/signup-user.dto';
import { PrismaService } from '../prisma/prisma.service';
import * as bcrypt from 'bcrypt';
import { ConfigService } from '@nestjs/config';
import { roles } from '@prisma/client';
import { DateTime } from 'luxon';
import { LoginUserDto } from './dto/login-user.dto';
import { JwtService } from '@nestjs/jwt';
import { ForgotPasswordDto } from './dto/forgot-password.dto';
import { ResetPasswordDto } from './dto/reset-password.dto';
import { EMAIL_TEMPLATE, EmailsService } from '../emails/emails.service';
import { UsersService } from '../users/users.service';
import { Users } from '../users/users.model';
import { ForgotPasswordResponseDto } from './dto/forgot-password-response.dto';
import { ResetPasswordResponseDto } from './dto/ResetPasswordResponse.dto';

// todo refactor all specific user related operations to user service.
@Injectable()
export class AuthService {
  constructor(
    private readonly prisma: PrismaService,
    private readonly configService: ConfigService,
    private jwtService: JwtService,
    private readonly emailsService: EmailsService,
    private readonly usersService: UsersService,
  ) {}

  /*
   @description method throws error if timestamps are in 24-hour period.
   @param timestamps an array of timestamps
   @param message customize the error message
   @param status customize the http status code
   */
  checkIfAttemptsInLast24h(
    timestamps: Date[],
    message: string = 'Too many attempts in 24 hour period.',
    status: HttpStatus = HttpStatus.FORBIDDEN,
  ): void {
    // account locked check
    if (timestamps && timestamps.length >= 3) {
      const aDayAgo = DateTime.now().minus({ days: 1 });

      const isIn24h = timestamps.reduce((accumulator, current) => {
        const value = DateTime.fromISO(current.toISOString());
        accumulator = accumulator === value >= aDayAgo; // basically an AND of whole array to check if is past 24 hours.
        return accumulator;
      }, true);

      if (isIn24h) {
        throw new HttpException(message, status);
      }
    }
  }

  async register(data: SignupUserDto): Promise<Users> {
    if (data.password !== data.repeat_password) {
      throw new HttpException('Passwords do not match', HttpStatus.BAD_REQUEST);
    }

    const duplicate = await this.prisma.users.count({
      where: { email: data.email },
    });
    if (duplicate > 0) {
      throw new HttpException('Email is already in use', HttpStatus.CONFLICT);
    }

    const hashed = await bcrypt.hash(data.password, 10);
    let role: roles = roles.customer;
    if (this.configService.get<string>('AUTO_ROLE') === 'TRUE') {
      const email = data.email.split('@');
      const suffix = email[0].split('+');
      if (suffix.length > 1) {
        if (suffix.at(-1) === 'admin') {
          role = roles.admin;
        } else if (suffix.at(-1) === 'manager') {
          role = roles.manager;
        }
      }
    }

    const user = this.prisma.users.create({
      data: {
        first_name: data.first_name,
        last_name: data.last_name,
        email: data.email,
        password: hashed,
        role,
      },
      select: {
        id: true,
        email: true,
        role: true,
        last_name: true,
        first_name: true,
        created_at: true,
      },
    });

    await this.emailsService.sendEmail(EMAIL_TEMPLATE.WELCOME, {
      user_first_name: data.first_name,
    });

    return user;
  }

  async recordFailedLoginAttempt(
    id: string,
    timestamps: Date[],
  ): Promise<Users> {
    const ts_length = timestamps.push(new Date());
    if (ts_length > 3) {
      timestamps.shift();
    }

    return this.prisma.users.update({
      where: { id: id },
      data: {
        failed_login_attempts: timestamps.length,
        failed_login_attempts_timestamps: timestamps,
      },
    });
  }

  async recordSuccessfulLoginAttempt(
    id: string,
    token: string,
  ): Promise<Users> {
    const now = new Date().toISOString();
    return this.prisma.users.update({
      where: { id: id },
      data: { login_at: now, session_token: token },
    });
  }

  async recordPasswordResetRequest(
    id: string,
    timestamps: Date[],
    resetToken: string,
  ): Promise<Users> {
    const ts_length = timestamps.push(new Date());
    if (ts_length > 3) {
      timestamps.shift();
    }

    return this.prisma.users.update({
      where: { id: id },
      data: {
        password_reset_requests: timestamps.length,
        password_reset_requests_timestamps: timestamps,
        password_reset_token: resetToken,
      },
    });
  }

  async forgotPassword(
    data: ForgotPasswordDto,
  ): Promise<ForgotPasswordResponseDto> {
    const user = await this.usersService.getUserByEmail(data.email);

    // too many reset attempts check
    this.checkIfAttemptsInLast24h(
      user.password_reset_requests_timestamps,
      'Too many password reset requests in a 24 hour period. Try again later.',
    );

    const token = this.jwtService.sign({
      user: user.id,
    });

    const forgotPasswordData = await this.recordPasswordResetRequest(
      user.id,
      user.password_reset_requests_timestamps,
      token,
    );

    await this.emailsService.sendEmail(EMAIL_TEMPLATE.FORGOT_PASSWORD, {
      reset_token: token,
    });

    return {
      message:
        'An email has been sent with a link to reset the password. Check your email.',
      request_count: forgotPasswordData.password_reset_requests,
      request_timestamps: forgotPasswordData.password_reset_requests_timestamps,
    };
  }

  async resetPasswordWithToken(
    id: string,
    new_password: string,
  ): Promise<Users> {
    const hashed = await bcrypt.hash(new_password, 10);

    return this.prisma.users.update({
      where: { id: id },
      data: {
        password: hashed,
        password_last_updated: new Date(),
        password_reset_token: null,
        // clear previous attempts
        password_reset_requests: 0,
        password_reset_requests_timestamps: [],
        // unlocks account
        failed_login_attempts: 0,
        failed_login_attempts_timestamps: [],
      },
    });
  }

  async logout(id: string): Promise<Users> {
    const now = new Date().toISOString();
    return this.prisma.users.update({
      where: { id: id },
      data: {
        session_token: null,
        logout_at: now,
      },
    });
  }

  async login(data: LoginUserDto): Promise<Users & { token: string }> {
    const user = await this.usersService.getUserByEmail(data.email);

    // account locked check
    this.checkIfAttemptsInLast24h(
      user.failed_login_attempts_timestamps,
      'Too many failed login attempts, account is locked. Try again later.',
    );
    const isCorrectPassword = await bcrypt.compare(
      data.password,
      user.password,
    );
    if (isCorrectPassword) {
      const token = await this.jwtService.signAsync({
        user: user.id,
        role: user.role,
      });

      await this.recordSuccessfulLoginAttempt(user.id, token);
      await this.emailsService.sendEmail(EMAIL_TEMPLATE.LOGIN_SUCCESSFUL);
      return {
        id: user.id,
        token,
        first_name: user.first_name,
        last_name: user.last_name,
        email: user.email,
        created_at: user.created_at,
        role: user.role,
        login_at: user.login_at,
        password_last_updated: user.password_last_updated,
      };
    } else {
      await this.recordFailedLoginAttempt(
        user.id,
        user.failed_login_attempts_timestamps,
      );
      await this.emailsService.sendEmail(EMAIL_TEMPLATE.LOGIN_FAIL);
      throw new HttpException(
        'Wrong password. Try again or use the forgot password api to reset it.',
        HttpStatus.UNAUTHORIZED,
      );
    }
  }

  async changePassword(
    data: ResetPasswordDto,
  ): Promise<ResetPasswordResponseDto> {
    if (data.password !== data.repeat_password) {
      throw new HttpException('Passwords do not match', HttpStatus.BAD_REQUEST);
    }
    let payload: { user: string };
    try {
      payload = await this.jwtService.verifyAsync(data.reset_token);
    } catch (error) {
      throw new HttpException(
        `An error occurred when validating reset token: ${error.message}`,
        HttpStatus.UNAUTHORIZED,
      );
    }
    const { password_reset_token, id } = await this.usersService.getUserById(
      payload.user,
    );

    if (password_reset_token === data.reset_token) {
      const { password_last_updated } = await this.resetPasswordWithToken(
        id,
        data.password,
      );
      return {
        message: 'The password has been reset successfully!',
        password_last_updated,
      };
    } else {
      throw new HttpException('The token is not valid.', HttpStatus.FORBIDDEN);
    }
  }
}
