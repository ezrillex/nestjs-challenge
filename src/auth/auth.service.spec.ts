import { Test, TestingModule } from '@nestjs/testing';
import { AuthService } from './auth.service';
import { ConfigService } from '@nestjs/config';
import { PrismaService } from '../prisma/prisma.service';
import { roles, Users } from '@prisma/client';
import { JwtModule, JwtService } from '@nestjs/jwt';
import { UsersService } from '../users/users.service';
import { EmailsService } from '../emails/emails.service';

jest.mock('@sendgrid/mail', () => {
  return {
    __esModule: true, // Indicates that this is a module with default export
    default: {
      setApiKey: jest.fn(), // Mock setApiKey method
      send: jest.fn(), // Mock send method
    },
  };
});

describe('AuthService', () => {
  let service: AuthService;
  let configService: ConfigService;
  let prismaService: PrismaService;
  let usersService: UsersService;
  let jwtService: JwtService;

  beforeEach(async () => {
    jest.clearAllMocks();
    const module: TestingModule = await Test.createTestingModule({
      providers: [
        PrismaService,
        ConfigService,
        AuthService,
        UsersService,
        EmailsService,
      ],
      imports: [
        JwtModule.registerAsync({
          global: false,
          useFactory: async () => ({
            secret: 'TEST_SECRET',
            signOptions: { expiresIn: '24h' },
          }),
        }),
      ],
    }).compile();

    service = module.get<AuthService>(AuthService);
    configService = module.get(ConfigService);
    prismaService = module.get(PrismaService);
    jwtService = module.get(JwtService);
    usersService = module.get(UsersService);
  });

  it('AuthService should be defined', () => {
    expect(service).toBeDefined();
  });

  it('ConfigService should be defined', () => {
    expect(configService).toBeDefined();
  });

  it('PrismaService should be defined', () => {
    expect(prismaService).toBeDefined();
  });

  it('UsersService should be defined', () => {
    expect(usersService).toBeDefined();
  });

  it('JwtService should be defined', () => {
    expect(jwtService).toBeDefined();
  });

  it('checkIfAttemptsInLast24h should be defined', () => {
    expect(service.checkIfAttemptsInLast24h).toBeDefined();
  });

  it('register should be defined', () => {
    expect(service.register).toBeDefined();
  });

  it('recordFailedLoginAttempt should be defined', () => {
    expect(service.recordFailedLoginAttempt).toBeDefined();
  });

  it('recordSuccessfulLoginAttempt should be defined', () => {
    expect(service.recordSuccessfulLoginAttempt).toBeDefined();
  });

  it('recordPasswordResetRequest should be defined', () => {
    expect(service.recordPasswordResetRequest).toBeDefined();
  });

  it('forgotPassword should be defined', () => {
    expect(service.forgotPassword).toBeDefined();
  });

  it('resetPasswordWithToken should be defined', () => {
    expect(service.resetPasswordWithToken).toBeDefined();
  });

  it('logout should be defined', () => {
    expect(service.logout).toBeDefined();
  });

  it('login should be defined', () => {
    expect(service.login).toBeDefined();
  });

  it('changePassword should be defined', () => {
    expect(service.changePassword).toBeDefined();
  });

  describe('checkIfAttemptsInLast24h', () => {
    it('does not throw an error when no timestamps are provided', () => {
      expect(() => service.checkIfAttemptsInLast24h([])).not.toThrow();
    });

    it('throws an error when more than two timestamps fall within a 24-hour period', () => {
      const now = new Date();
      expect(() =>
        service.checkIfAttemptsInLast24h([now, now, now]),
      ).toThrowErrorMatchingSnapshot('too many in 24 h');
    });

    it('does not throw an error when exactly two timestamps fall within a 24-hour period', () => {
      const now = new Date();
      expect(() => service.checkIfAttemptsInLast24h([now, now])).not.toThrow();
    });

    it('does not throw an error when input is undefined', () => {
      expect(() => service.checkIfAttemptsInLast24h(undefined)).not.toThrow();
    });

    it('does not throw an error when all timestamps are in the past', () => {
      const past = new Date(1999, 2, 20, 1, 50, 20);
      expect(() =>
        service.checkIfAttemptsInLast24h([past, past, past]),
      ).not.toThrow();
    });
  });

  describe('registerUser', () => {
    it('throws an error when password and repeat_password do not match', async () => {
      await expect(
        service.register({
          email: 'some@email.com',
          password: 'somepassword!',
          last_name: 'smith',
          first_name: 'john',
          repeat_password: 'somepassword',
        }),
      ).rejects.toThrowErrorMatchingSnapshot('password mismatch');
    });

    it('throws an error when email is already registered', async () => {
      jest.spyOn(prismaService.users, 'count').mockResolvedValue(1);

      await expect(
        service.register({
          email: 'some@email.com',
          password: 'somepassword',
          last_name: 'smith',
          first_name: 'john',
          repeat_password: 'somepassword',
        }),
      ).rejects.toThrowErrorMatchingSnapshot('email is already in use');
    });

    it('successfully creates a user when all validations pass', async () => {
      let result;
      jest
        .spyOn(prismaService.users, 'create')
        .mockImplementation(({ data }) => {
          result = data;
          return null;
        });

      await service.register({
        email: 'some@email.com',
        password: 'somepassword',
        last_name: 'smith',
        first_name: 'john',
        repeat_password: 'somepassword',
      });

      expect(result).toEqual({
        email: 'some@email.com',
        password: expect.any(String),
        last_name: 'smith',
        first_name: 'john',
        role: roles.customer,
      });
    });

    it('assigns role as customer even when email ends with +admin when auto-role is disabled', async () => {
      let result;
      jest
        .spyOn(prismaService.users, 'create')
        .mockImplementation(({ data }) => {
          result = data;
          return null;
        });
      jest.spyOn(configService, 'get').mockImplementation(() => {
        return 'FALSE';
      });

      await service.register({
        email: 'name+admin@email.com',
        password: 'somepassword',
        last_name: 'smith',
        first_name: 'john',
        repeat_password: 'somepassword',
      });

      expect(result).toEqual({
        email: 'name+admin@email.com',
        password: expect.any(String),
        last_name: 'smith',
        first_name: 'john',
        role: roles.customer,
      });
    });

    it('assigns role as admin when email ends with +admin and auto-role is enabled', async () => {
      let result;
      jest
        .spyOn(prismaService.users, 'create')
        .mockImplementation(({ data }) => {
          result = data;
          return null;
        });

      jest.spyOn(configService, 'get').mockReturnValue('TRUE');

      await service.register({
        email: 'name+admin@email.com',
        password: 'somepassword',
        last_name: 'smith',
        first_name: 'john',
        repeat_password: 'somepassword',
      });

      expect(result).toEqual({
        email: 'name+admin@email.com',
        password: expect.any(String),
        last_name: 'smith',
        first_name: 'john',
        role: roles.admin,
      });
    });

    it('assigns role as manager when email ends with +manager and auto-role is enabled', async () => {
      let result;
      jest
        .spyOn(prismaService.users, 'create')
        .mockImplementation(({ data }) => {
          result = data;
          return null;
        });

      jest.spyOn(configService, 'get').mockReturnValue('TRUE');

      await service.register({
        email: 'name+manager@email.com',
        password: 'somepassword',
        last_name: 'smith',
        first_name: 'john',
        repeat_password: 'somepassword',
      });

      expect(result).toEqual({
        email: 'name+manager@email.com',
        password: expect.any(String),
        last_name: 'smith',
        first_name: 'john',
        role: roles.manager,
      });
    });
  });

  describe('recordFailedLoginAttempt', () => {
    it('updates the database to append the current timestamp and retain only the latest three timestamps', async () => {
      let result;

      jest
        .spyOn(prismaService.users, 'update')
        .mockImplementation(({ data }) => {
          result = data;
          return null;
        });
      const past = new Date(1999, 2, 20, 1, 50, 20);
      const now = new Date(2020, 12, 30, 1, 45, 50);
      jest.useFakeTimers().setSystemTime(now);

      await service.recordFailedLoginAttempt('some id', [past, past, past]);
      expect(result.failed_login_attempts_timestamps).toEqual([
        past,
        past,
        now,
      ]);
    });
  });

  describe('recordSuccessfulLoginAttempt', () => {
    it('sends a correctly structured update query to the database with the current timestamp and session token', async () => {
      let result;
      jest.spyOn(prismaService.users, 'update').mockImplementation((data) => {
        result = data;
        return null;
      });
      const now = new Date();
      jest.useFakeTimers().setSystemTime(now);

      await service.recordSuccessfulLoginAttempt('test_id', 'test_token');
      expect(result).toEqual({
        where: { id: 'test_id' },
        data: {
          login_at: now.toISOString(),
          session_token: 'test_token',
        },
      });
    });
  });

  describe('forgotPasswordRequest', () => {
    it('updates the database to append a new timestamp and retain only the latest three timestamps', async () => {
      let result;

      jest.spyOn(prismaService.users, 'update').mockImplementation((data) => {
        result = data;
        return null;
      });
      const past = new Date(1999, 2, 20, 1, 50, 20);
      const now = new Date(2020, 12, 30, 1, 45, 50);
      jest.useFakeTimers().setSystemTime(now);

      await service.recordPasswordResetRequest(
        'some id',
        [past, past, past],
        'made up token',
      );
      expect(result).toMatchSnapshot('expected database update');
    });
  });

  describe('resetPassword', () => {
    it('updates the database with the new password', async () => {
      let result;

      jest.spyOn(prismaService.users, 'update').mockImplementation((data) => {
        data.data.password = expect.any(String);
        result = data;
        return null;
      });
      const now = new Date(2020, 12, 30, 1, 45, 50);
      jest.useFakeTimers().setSystemTime(now);

      await service.resetPasswordWithToken('some id', 'the new password');
      expect(result).toMatchSnapshot('the expected database update');
    });

    it('ensures the password is hashed before being saved in the database', async () => {
      let result;

      jest.spyOn(prismaService.users, 'update').mockImplementation((data) => {
        result = data;
        return null;
      });
      const now = new Date(2020, 12, 30, 1, 45, 50);
      jest.useFakeTimers().setSystemTime(now);

      await service.resetPasswordWithToken('some id', 'the new password');
      expect(result.data.password).not.toEqual('the new password');
    });
  });

  describe('logoutUser', () => {
    it('updates the database with a valid logout request for the user', async () => {
      let result;

      jest.spyOn(prismaService.users, 'update').mockImplementation((data) => {
        result = data;
        return null;
      });
      const now = new Date(2020, 12, 30, 1, 45, 50);
      jest.useFakeTimers().setSystemTime(now);

      await service.logout('some id');
      expect(result).toMatchSnapshot('the expected database update');
    });
  });

  describe('loginUser', () => {
    it('throws an error when the email does not exist in the database', async () => {
      jest.spyOn(prismaService.users, 'findUnique').mockReset();
      await expect(
        service.login({
          email: 'some@email.com',
          password: 'some password',
        }),
      ).rejects.toThrowErrorMatchingSnapshot('email not found response');
    });

    it('throws an error when the password is incorrect', async () => {
      // util is tested already.
      const util = jest
        .spyOn(service, 'checkIfAttemptsInLast24h')
        .mockImplementation(() => {});

      // tested already
      jest.spyOn(service, 'recordFailedLoginAttempt').mockResolvedValue(null);

      // hash of 'therightpassword'
      jest.spyOn(usersService, 'findOneByEmail').mockResolvedValue({
        id: 'test id',
        password:
          '$2y$10$Xzx25dxZ/cq0xn7toLj1HuZ1ZMfId8gRuR3VBGvX9fWdboh9xZrMa',
      } as Users);

      await expect(
        service.login({
          email: 'some@email.com',
          password: 'thewrongpassword',
        }),
      ).rejects.toThrowErrorMatchingSnapshot('wrong password response');
      expect(util).toHaveBeenCalled();
    });

    it('returns a token and role when credentials are valid', async () => {
      // is already tested.
      jest
        .spyOn(service, 'checkIfAttemptsInLast24h')
        .mockImplementation(() => {});
      jest.spyOn(service, 'recordFailedLoginAttempt').mockResolvedValue(null);
      jest
        .spyOn(service, 'recordSuccessfulLoginAttempt')
        .mockResolvedValue(null);
      jest.spyOn(jwtService, 'signAsync').mockResolvedValue('my_token');

      jest.spyOn(usersService, 'findOneByEmail').mockResolvedValue({
        id: 'test id',
        role: roles.customer,
        password:
          '$2b$10$fhpIpbwfaJlbiOZPDKlKxuV71lFjOk3EeX9npe.LSF0fIyiouR2i2',
      } as Users);

      await expect(
        service.login({
          email: 'some@email.com',
          password: 'therightpassword',
        }),
      ).resolves.toMatchSnapshot('ok response');
    });
  });

  describe('forgotPassword', () => {
    it('returns a status message with a reset token', async () => {
      // is already tested.
      jest
        .spyOn(service, 'checkIfAttemptsInLast24h')
        .mockImplementation(() => {});
      jest.spyOn(service, 'recordPasswordResetRequest').mockResolvedValue({
        password_reset_requests: 1,
        password_reset_requests_timestamps: [
          new Date(2020, 12, 12, 12, 12, 12),
        ],
      } as Users);
      jest.spyOn(jwtService, 'sign').mockReturnValue('my_token');

      jest.spyOn(usersService, 'findOneByEmail').mockResolvedValue({} as Users);

      await expect(
        service.forgotPassword({
          email: 'some@email.com',
        }),
      ).resolves.toEqual({
        message:
          'An email has been sent with a link to reset the password. Check your email.',
        request_count: 1,
        request_timestamps: [new Date(2020, 12, 12, 12, 12, 12)],
      });
    });
  });

  describe('changePassword', () => {
    it('throws an error when the password and repeat password do not match', async () => {
      await expect(
        service.changePassword({
          password: 'a',
          repeat_password: 'b',
          reset_token: 'my_token',
        }),
      ).rejects.toThrowErrorMatchingSnapshot('mismatch password response');
    });

    it('throws an error when the token is malformed', async () => {
      await expect(
        service.changePassword({
          password: 'a',
          repeat_password: 'a',
          reset_token: 'my_token',
        }),
      ).rejects.toThrowErrorMatchingSnapshot('token is malformed');
    });

    it('throws an error when the token is invalid', async () => {
      await expect(
        service.changePassword({
          password: 'a',
          repeat_password: 'a',
          reset_token:
            'eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJPbmxpbmUgSldUIEJ1aWxkZXIiLCJpYXQiOjE3MzU5NDAyMzYsImV4cCI6MTc2NzQ3NjIzNiwiYXVkIjoid3d3LmV4YW1wbGUuY29tIiwic3ViIjoianJvY2tldEBleGFtcGxlLmNvbSIsIkdpdmVuTmFtZSI6IkpvaG5ueSIsIlN1cm5hbWUiOiJSb2NrZXQiLCJFbWFpbCI6Impyb2NrZXRAZXhhbXBsZS5jb20iLCJSb2xlIjpbIk1hbmFnZXIiLCJQcm9qZWN0IEFkbWluaXN0cmF0b3IiXX0.vwn02EWxfKdm0bVyaLFQVpTWevnsEah5lRu_gU27vCk',
        }),
      ).rejects.toThrowErrorMatchingSnapshot('token is invalid');
    });

    it("throws an error when the token payload does not match the user's current token (e.g., token reused or invalidated)", async () => {
      jest
        .spyOn(jwtService, 'verifyAsync')
        .mockResolvedValue({ user: 'some_user_id' });
      jest.spyOn(usersService, 'findOneByID').mockResolvedValue({
        password_reset_token: 'some_random_token',
      } as Users);

      await expect(
        service.changePassword({
          password: 'a',
          repeat_password: 'a',
          reset_token: 'my_token',
        }),
      ).rejects.toThrowErrorMatchingSnapshot('token mismatch, invalidated');
    });

    it('returns a success message when the password is successfully reset', async () => {
      // is already tested.
      jest
        .spyOn(jwtService, 'verifyAsync')
        .mockResolvedValue({ user: 'some_user_id' });
      jest.spyOn(usersService, 'findOneByID').mockResolvedValue({
        password_reset_token: 'my_token',
      } as Users);
      jest.spyOn(service, 'resetPasswordWithToken').mockResolvedValue({
        password_last_updated: new Date(2020, 12, 12, 12, 12, 12),
      } as Users);

      await expect(
        service.changePassword({
          password: 'a',
          repeat_password: 'a',
          reset_token: 'my_token',
        }),
      ).resolves.toEqual({
        message: 'The password has been reset successfully!',
        password_last_updated: new Date(2020, 12, 12, 12, 12, 12),
      });
    });
  });
});
