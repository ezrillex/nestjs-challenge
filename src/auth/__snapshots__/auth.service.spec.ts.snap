// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AuthService change password  should throw if password mismatch: mismatch password response 1`] = `"Passwords do not match"`;

exports[`AuthService change password  should throw if payload user token mismatch (requested again, or already used and tried to use link again): token mismatch, invalidated 1`] = `"The token is not valid."`;

exports[`AuthService change password  should throw if token is invalid: token is invalid 1`] = `"An error occurred when validating reset token: invalid signature"`;

exports[`AuthService change password  should throw if token is malformed: token is malformed 1`] = `"An error occurred when validating reset token: jwt malformed"`;

exports[`AuthService create user should error if email is already in use: email is already in use 1`] = `"Email is already in use"`;

exports[`AuthService create user should error if password is different: password mismatch 1`] = `"Passwords do not match"`;

exports[`AuthService find one by ID user should error if id was not found: user with ID not found 1`] = `"Couldn't find an account with the associated id ."`;

exports[`AuthService find one by email user should error if email was not found: user with email not found 1`] = `"Couldn't find your account. Make sure this is the right email."`;

exports[`AuthService forgot password requested should modify the db to append timestamp and shift it to 3: expected database update 1`] = `
{
  "data": {
    "password_reset_requests": 3,
    "password_reset_requests_timestamps": [
      1999-03-20T07:50:20.000Z,
      1999-03-20T07:50:20.000Z,
      2021-01-30T07:45:50.000Z,
    ],
    "password_reset_token": "made up token",
  },
  "where": {
    "id": "some id",
  },
}
`;

exports[`AuthService is in 24 hour period utility Should error if 3 timestamps are in 24 hours: too many in 24 h 1`] = `"Too many attempts in 24 hour period."`;

exports[`AuthService login user should error if credentials email doesnt exist: email not found response 1`] = `"Couldn't find your account. Make sure this is the right email."`;

exports[`AuthService login user should error if password is incorrect: wrong password response 1`] = `"Wrong password. Try again or use the forgot password api to reset it."`;

exports[`AuthService login user should return token and role if all ok: ok response 1`] = `
{
  "role": "customer",
  "token": "my_token",
}
`;

exports[`AuthService logout user should update db with valid logout user request: the expected database update 1`] = `
{
  "data": {
    "logout_at": "2021-01-30T07:45:50.000Z",
    "session_token": null,
  },
  "where": {
    "id": "some id",
  },
}
`;

exports[`AuthService reset password should update the database: the expected database update 1`] = `
{
  "data": {
    "failed_login_attempts": 0,
    "failed_login_attempts_timestamps": [],
    "password": Any<String>,
    "password_last_updated": 2021-01-30T07:45:50.000Z,
    "password_reset_requests": 0,
    "password_reset_requests_timestamps": [],
    "password_reset_token": null,
  },
  "where": {
    "id": "some id",
  },
}
`;
